### DISCUSSION TITLE
Looking for some help for making a watertight mesh

### DISCUSSION THREAD

**Discussion by conorssmith** [pyvista/pyvista#1589] [Q&A] [answered]
Hi there,

I have spent the better part of two weeks trying to create watertight meshes in Python from organ contour data. My goal is to make watertight meshes so I can calculate the volume of intersection between two meshes. The problem is that increasing alpha loses the surface shape and creates a concave hull. 

Below I have attached a picture showing one of my attempts, the mesh is riddled with holes, 

I'm a bit of a newb so I'd really appreciate any help!

![PyVista Mesh](https://user-images.githubusercontent.com/88810617/129299938-cd44c256-99dd-4478-95e3-53b3bd31bed9.JPG)

```
import numpy as np
import pyvista as pv
import pymeshfix

point_cloud = np.genfromtxt('ct_prostate_contour_data.csv', delimiter=',')
cloud = pv.PolyData(point_cloud)
mesh = cloud.delaunay_3d(alpha=2).extract_geometry().clean(point_merging=True)
mesh.extract_largest().clean()
mesh.compute_normals(inplace=True)
mesh.triangulate()
mesh.extract_largest()
meshfix = pymeshfix.MeshFix(mesh)
meshfix.repair()
mesh.plot(show_edges=True)
```

[ct_prostate_contour_data.csv](https://github.com/pyvista/pyvista/files/6979659/ct_prostate_contour_data.csv)
[ct_rectum_contour_data.csv](https://github.com/pyvista/pyvista/files/6979661/ct_rectum_contour_data.csv)
[ct_bladder_contour_data.csv](https://github.com/pyvista/pyvista/files/6979662/ct_bladder_contour_data.csv)

**Comment by akaszynski** [comment, marked as answer]
This is indeed a hard one.  VTK's delaunay filter won't be able to work in this case, and I think your best bet it to work on a custom algorithm to "connect the dots" and construct the triangles yourself.

I think it's possible if you extract the individual sections and write something to generate triangles between the sections (perhaps using some nearest neighbor on the other section).  Here's something basic to extract the sections for `ct_prostate_contour_data.csv`

```py
import numpy as np
import pyvista as pv
import pymeshfix

point_cloud = np.genfromtxt('ct_prostate_contour_data.csv', delimiter=',')
cloud = pv.PolyData(point_cloud)

height = cloud.bounds[-1] - cloud.bounds[-2]
zmin = cloud.bounds[-2]
sec_height = height/19 # 2 sections at a time

z = cloud.points[:, 2]
tol = 0.1
i = 1
mask_below = z < zmin + sec_height*(i + 2) + tol
mask_above = z > zmin + sec_height*i - tol
mask = np.logical_and(mask_above, mask_below)

sections = cloud.extract_points(mask)
```

![tmp](https://user-images.githubusercontent.com/11981631/129306780-6c4f81bc-b367-4aaa-ba89-4a5015afc059.png)

**Reply by conorssmith** [reply]
Perhaps I just haven't delved deeply enough into the problem you've suggested but wouldn't each contour slice require the same number of data points to ensure a closed surface? (Assuming triangles only) 

Also I'm not sure I'm a good enough programmer to handle the slices with more than one section such as the top slice of the prostate data. 😅

![Point Cloud](https://user-images.githubusercontent.com/88810617/129310408-fc700bd9-5e66-4338-9320-bbcca21e2c73.JPG)

**Reply by akaszynski** [reply]
More progress...
```py
from pykdtree.kdtree import KDTree
import numpy as np
import pyvista as pv
import pymeshfix

point_cloud = np.genfromtxt('ct_prostate_contour_data.csv', delimiter=',')
cloud = pv.PolyData(point_cloud)

height = cloud.bounds[-1] - cloud.bounds[-2]
zmin = cloud.bounds[-2]
sec_height = height/19 # 2 sections at a time

z = cloud.points[:, 2]
tol = 0.1


def parse_section(i):
    mask_below = z < zmin + sec_height*(i + 1) + tol
    mask_above = z > zmin + sec_height*i - tol
    mask = np.logical_and(mask_above, mask_below)
    section = cloud.extract_points(mask)
    # section['section number'] = i
    return section


sections = [parse_section(i) for i in range(19)]


def triangles_from_sections(section_a, section_b):
    """Return triangles from two sections using a kdtree"""
    section_tree = KDTree(section_b.points)
    dist, matches = section_tree.query(section_a.points, 3)

    # Must use indices of the original IDs
    triangle_a = section_a['vtkOriginalPointIds']
    triangle_bcd = section_b['vtkOriginalPointIds'][matches]
    # breakpoint()
    triangle_bc = triangle_bcd[:, :2]
    triangle_bd = triangle_bcd[:, [0, 2]]

    padding = np.ones((section_a.n_points, 1), dtype=triangle_a.dtype)*3
    triangles_a = np.hstack((padding,
                             triangle_a.reshape(-1, 1),
                             triangle_bc))

    triangles_b = np.hstack((padding,
                             triangle_a.reshape(-1, 1),
                             triangle_bd))

    return np.vstack((triangles_a, triangles_b))

triangles = []
for i in range(18):
    section_a = sections[i]
    section_b = sections[i + 1]
    triangles.append(triangles_from_sections(section_a, section_b))
    triangles.append(triangles_from_sections(section_b, section_a))


triangles = np.vstack(triangles)

cloud.faces = triangles
areas = cloud.compute_cell_sizes(area=True)['Area']

mesh = sizes.extract_cells(areas < 15).extract_surface().triangulate()
mesh.plot(color='w')
```
![tmp](https://user-images.githubusercontent.com/11981631/129387797-a83b46b6-e759-41c4-85e8-a9a1168ddd1a.png)

**Comment by darikg** [comment]
Reconstructing a surface from serial planar contours is an old problem; see e.g. *Approximating complex surfaces by triangulation of contour lines* Keppel 1977 and *Optimal Surface Reconstruction from Planar Contours* Fuchs et al 1977. Those early papers dealt with a single "tube" and subsequent work generalized to arbitary branching (which it looks like is occurring in your dataset?) and merging of contours. While there's tons of literature on the subject, I'm not aware of any publically available implementations, although I'd be delighted to be wrong there. I have a poorly-tested implementation of some of that early work I could throw up on github, but only for the simple tube case, no branching/merging.

**Comment by eyildiz-ugoe** [comment]
I am trying to do the same with my meshes. They are all non-watertight and I'd like to try pyvista to fix them. However, I cannot load them as you do as I don't use a .csv file for my point clouds. They are generated via open3d library and thus pyvista's function throws an error, naturally, since they are different libraries. My question is: is there a generic way to fix this? One usually does not use .csv file to keep PCD data and a way to handle this situation would be really nice. Many of us are using other libraries (e.g. open3d, trimesh) to do various operations. Fixing a non-watertight mesh, however, is not possible using those. How could we try pyvista in that regard? How can we load a pointcloud that was generated via a different library? Here is an example:

P.S: I am aware why this throws an error, it's very obvious. I am interested in a solution to use both libraries. Pyvista seems to offer a fix for non-watertight meshes and I'd like to try that on my meshes.

        ```
        # using open3d, get the point cloud ready
        pcd_info        = PlyData.read(path_to_point_cloud)
        pcd_vertexses   = pcd_info['vertex'].data

        normals = np.stack([ pcd_vertexses[t] for t in ['nx', 'ny', 'nz']], axis=-1)
        colors  = np.stack([(pcd_vertexses[t] / 255.0) for t in ['red', 'green', 'blue']], axis=-1)
        points  = np.stack([ pcd_vertexses[t] for t in ('x', 'y', 'z'   )], axis=-1)

        point_cloud         = o3d.geometry.PointCloud()
        point_cloud.normals = o3d.utility.Vector3dVector(normals)
        point_cloud.colors  = o3d.utility.Vector3dVector(colors )
        point_cloud.points  = o3d.utility.Vector3dVector(points )

        # now use pyvista to fix it
        surf = point_cloud.reconstruct_surface(nbr_sz=20, sample_spacing=2)

        mf = pymeshfix.MeshFix(surf)
        mf.repair()
        repaired = mf.mesh

        pl = pv.Plotter()
        pl.add_mesh(point_cloud, color='k', point_size=10)
        pl.add_mesh(repaired)
        pl.add_title('Reconstructed Surface')
        pl.show()
```

Output: `AttributeError: 'open3d.cuda.pybind.geometry.PointCloud' object has no attribute 'reconstruct_surface`

**Comment by adam-grant-hendry** [comment]
> They are generated via open3d library and thus pyvista's function throws an error, naturally, since they are different libraries. My question is: is there a generic way to fix this?

@eyildiz-ugoe Yes. Allow me to demonstrate below.

@conorssmith @darikg @eino @akaszynski I **highly** recommend using the [`open3d`](http://www.open3d.org/docs/release/tutorial/geometry/surface_reconstruction.html) library in conjunction with either `pyvista` or `polyscope` (visualization only) for this problem.

To date, there are three well-known algorithms for reconstructing surfaces from point clouds:

- Alpha Shapes (1983)
- Ball Pivot (1999)
- Poisson (2006)

Ball Pivot and Poisson are vastly improved over Alpha Shapes as they use point normal information to distinguish between "inner vs. outer surfaces" (alpha shapes suffer from this). You can estimate point normals from point clouds by fitting planes to the points. `open3d` provides functions for estimating normals in addition to creating surfaces using the above algorithms.

To use `open3d` with `pyvista`, you need to learn how to extract relevant data from the `open3d` outputs.

**Pro Tip**: use an [`ipython`](https://ipython.readthedocs.io/en/stable/index.html) terminal, or similar, to get intellisense and real-time feedback on acceptable method parameters to use.

```python
import numpy as np
import open3d as o3d
import pyvista as pv


if __name__ == '__main__':
    points = np.genfromtxt('ct_prostate_contour_data.csv', delimiter=',')

    cloud = o3d.geometry.PointCloud()
    cloud.points = o3d.utility.Vector3dVector(points)
    cloud.estimate_normals()

    # Before Fixing Normals
    o3d.visualization.draw_geometries([cloud,], point_show_normal=True)

    # Fix Normals
    cloud.orient_normals_consistent_tangent_plane(25)

    # After Fixing Normals
    o3d.visualization.draw_geometries([cloud,], point_show_normal=True)

    plotter = pv.Plotter()

    trimesh, densities = o3d.geometry.TriangleMesh.create_from_point_cloud_poisson(cloud)

    v = np.asarray(trimesh.vertices)
    f = np.array(trimesh.triangles)
    f = np.c_[np.full(len(f), 3), f]

    envelope = pv.PolyData(v, f).clean()
    envelope.actor = plotter.add_mesh(envelope, color='red', show_edges=True, opacity=0.5)

    plotter.show()
```

### Fixing Normals
![image](https://user-images.githubusercontent.com/59346180/200071940-011ecd35-3706-4c8f-9e07-d1f506e5feae.png)

### Final Result
![image](https://user-images.githubusercontent.com/59346180/200071876-066bf4b8-edaf-43d2-ad56-b6718c3ec2c1.png)

If you want to close off the left "ear", you can use `pymeshfix` to close holes:

```python
import pymeshfix as mf

meshfix = mf.MeshFix(envelope)

# Repair also fills holes
meshfix.repair(verbose=True)

envelope = meshfix.mesh.clean().triangulate()
```

### Clean Final Mesh

https://user-images.githubusercontent.com/59346180/200073182-26051b7a-6850-46f8-a7a7-78349b9c7481.mp4

**Reply by adeak** [reply]
This comes up a lot, it would be nice if we could have something better for this in VTK or PyVista... Do you have any experience with [`vtkSurfaceReconstructionFilter`](https://vtk.org/doc/nightly/html/classvtkSurfaceReconstructionFilter.html)? Sounds like it's yet another method, and it also needs proper normals.

**Reply by adam-grant-hendry** [reply]
Normal estimation is key, yes. MeshLab uses VTK under the hood and [`pymeshlab`](https://pymeshlab.readthedocs.io/en/latest/) is available. I know they have these algorithms, either out-of-the-box or through 3rd party plugins. I don't have experience with `vtkSurfaceReconstructionFilter`, but I would use Poisson or Ball Pivot as those are widely recognized and established algorithms. I would recommend taking a look at  [`pymeshlab`](https://pymeshlab.readthedocs.io/en/latest/) to see if they've done it.

**Reply by adeak** [reply]
Thanks. I don't think we should do this ourselves unless VTK already has something we can expose.

**Reply by adam-grant-hendry** [reply]
Agreed.

**Reply by conorssmith** [reply]
What I ended up doing was using matplotlibs path function to create polygons of each contour (one for each image slice), path has a built-in points in polygon function which you can feed in a a grid of points all at once. The end result is a boolean array with ones/true inside the contour and zeros/false outside the contour, stacking these makes a 3D boolean mask which can be fed into a marching cubes algorithm. The meshes are closed but ended up not being accurate enough for my project, but I did use the boolean masks, you can set the grid spacing to a desired resolution for an accuracy/compute time tradeoff.

**Reply by adam-grant-hendry** [reply]
@adeak I did happen across [this](https://www.vtkjournal.org/browse/publication/718). Perhaps we could expose it via `cython`? What do you think?

**Reply by adeak** [reply]
Using cython (mostly as a tool for internal speed-up) has come up before, and we generally agreed not to introduce it for fear of greatly reducing maintainability and increasing the barrier for entry. I would think that this is even more so true for using it for external modules.

**Comment by adam-grant-hendry** [comment]
@adeak Starting this thread again here. There is also [PoissonExtractSurface](https://kitware.github.io/vtk-examples/site/Cxx/Points/PoissonExtractSurface/), so we might actually be able to expose this in `pyvista`. Let me know and I'll submit a PR.

**Reply by adeak** [reply]
I saw that earlier, but it seemed to me that it's an external module that one has to compile into VTK, which would probably make it a no-go for PyVista. I would expect the vast, vast majority of our users to use VTK wheels from PyPI. Or am I missing what's needed here?

**Reply by adam-grant-hendry** [reply]
You can use `pybind11` with `CMake`. `pybind11` works with `setuptools` so you can build from a `CMakeLists.txt`. That’s how `pygalmesh` does it.

I also put `pyembree` together with cython (you can see my repo). Might make sense to keep C++ or cython things in a separate repo though, like “pyvista-extensions”.
